### PR TITLE
Remove C/R from OPAL, OMPI

### PR DESCRIPTION
**Pull request by bosilca**
A second approach, more radical, for #7437.

A similar patch has been proposed in PRRTE (https://github.com/openpmix/prrte/pull/710)

### CODE REVIEW

**Comment by ibm-ompi** [comment]
The IBM CI (GNU/Scale) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/8373f5bcacf96e89efec4e940ddc392c

**Comment by ibm-ompi** [comment]
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/077215052112a3d2c47e10ea62adc10e

**Comment by ibm-ompi** [comment]
The IBM CI (PGI) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/4f9fcc5b64df9327ff1ae7bd2c4d12df

**Comment by jjhursey** [comment]
The diff is huge. How is this different than PR #8311 ?

**Comment by jjhursey** [comment]
I'm fine with removing the ft_event from the component structure, but it requires a broader discussion with the community since it impacts components in the open-source tree and those that are not in the open-source tree.

Why did you open a PR that duplicates the work of PR #8311 instead of just making a comment on that PR? I would have been happy to fix.

**Comment by bosilca** [comment]
I had this on my tree for a long time, I updated it early in the winter break before I was made aware of your work on #8311. I think they address the same issue, but [and this after a quick look at your PR] mine is cutting a little deeper in C/R support, with the goal of not leaving anything related to C/R behind.

I was indeed wondering how to address the ft_event, and I could not find a good path forward.

1. we **leave** the callback as is for the few components that have it, but we need to be clear that it will be hard to repurpose and even harder to remove in the future, in addition to confusing any developers working around these areas.
2. we **remove** the callback in which case, as you noticed we are breaking the ABI compatibility for few of our internal components, potentially disrupting vendor and open source components we don't officially ship. It's cleaner, no left-overs but potentially disruptive to existing installations.
3. we do **2**, and then use the MCA versioning for the few frameworks affected, allowing them to convert from the prior to current MCA base version. I was planning to do a proof of concept of this, but then I saw your PR and stopped there. I would be willing to take a stab at it, if this is the way we decide to move forward.

**Comment by jjhursey** [comment]
We should decide on which branch to collaborate on - so we don't duplicate effort. If this branch already has the ft_event removed along with all of the other bits then we can work here. I'll close my PR, and may try to take a look at why you are failing CI later today (don't wait on me though).

As for the `ft_event`, I think option 3 is the best path forward. We remove `ft_event` since it's probably not going to be re-purposed and thus dead code (and confusing to component implementers). Then increment the component major number in the frameworks impacted. That should help limit the impact on downstream components (before they are updated - which shouldn't be difficult to do).

**Comment by jjhursey** [comment]
This should fix one of the CI build issues:
```diff
diff --git a/oshmem/mca/spml/ucx/spml_ucx.c b/oshmem/mca/spml/ucx/spml_ucx.c
index 2c1ae190..1c226c1f 100644
--- a/oshmem/mca/spml/ucx/spml_ucx.c
+++ b/oshmem/mca/spml/ucx/spml_ucx.c
@@ -21,6 +21,7 @@
 #include "oshmem_config.h"
 #include "opal/datatype/opal_convertor.h"
 #include "opal/mca/common/ucx/common_ucx.h"
+#include "opal/util/opal_environ.h"
 #include "ompi/datatype/ompi_datatype.h"
 #include "ompi/mca/pml/pml.h"
```

The others seem to revolve around the missing `mca_pml_ob1_com_btl_comp` symbol.

**Comment by bosilca** [comment]
@jjhursey thanks for the patch and the hint on the root issue for the CI failures. I've been a little too eager to cut things out, and I removed the mca_pml_ob1_com_btl_comp function.

I agree with you the option 3 seems the most sensible path forward.

**Comment by ibm-ompi** [comment]
The IBM CI (PGI) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/d9c7a62a77c197606d098da72034e96b

**Comment by jjhursey** [comment]
The PGI CI failure is not related to this PR. It'll be fixed by PR #8325 when it gets merged.

**Comment by jjhursey** [comment]
bot:ibm:pgi:retest
/azp run

**Comment by jjhursey** [comment]
What's left to do here?

**Comment by bosilca** [comment]
Making sure we don't load components from a prior install (components that will have a different MCA version).

**Comment by jjhursey** [comment]
Any chance of getting this in before branching for v5?

**Comment by gpaulsen** [comment]
@bosilca please rebase.

**Review by jjhursey — CHANGES_REQUESTED** [review]
Overall this is looking good. I found a couple of notes.

The `crs` framework was deleted here. Additionally, the `snapc` framework was removed with ORTE. Some of the platform files still refer to these frameworks. It might be good to scrub them of that artifact as well.

File: opal/mca/mpool/mpool.h [hunk 1]
```diff
@@ -135,7 +123,6 @@ struct mca_mpool_base_module_t {
     mca_mpool_base_module_free_fn_t mpool_free;          /**< free function */
 
     mca_mpool_base_module_finalize_fn_t mpool_finalize;  /**< finalize */
-    mca_mpool_base_module_ft_event_fn_t mpool_ft_event;  /**< ft_event */
```
[the remark this answers is not in the document]
Comment (bosilca) [reply opal/mca/mpool/mpool.h:138]: We have no versioning for the mpool modules. I can solve this by increasing the version on the component, but this is getting ugly.

**Comment by bosilca** [comment]
The platform files are old and based on their content mostly unused. I tried to clean them based on what I remember as being removed, but I'm sure they deserve a little bit more attention.

**Comment by ibm-ompi** [comment]
The IBM CI (GNU/Scale) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/c4b0eae6e9490ff29da23f684dd7050b

**Comment by ibm-ompi** [comment]
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/8fa94f27d45b39a0796a570e4d3b867b

**Comment by ibm-ompi** [comment]
The IBM CI (PGI) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/5d5395192bbf1b9ae9117c7d5a6a05b6

**Comment by ibm-ompi** [comment]
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/a33ab0fff132c455ad05b65924adce0e

**Comment by ibm-ompi** [comment]
The IBM CI (PGI) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/bf7f2c243787e0b7579bc50bc6d417dc

**Comment by ibm-ompi** [comment]
The IBM CI (GNU/Scale) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/9af296a12b16be3328e50a53a91a8286

**Comment by bosilca** [comment]
@jjhursey all your comments have been addressed. I also squashed the entire commit into a single one. From my perspective, this PR is ready to go.

**Review by jjhursey — APPROVED** [review]
Looks good to me!

**Comment by jjhursey** [comment]
Thanks for everything Checkpoint/Restart functionality! Sad to see you go.

**Pull request merged**